### PR TITLE
Update flowmatch loss code to latest changes in the code base

### PR DESCRIPTION
**Pull request by alexhernandezgarcia**
Note: this PR includes the fixes in [PR 79](https://github.com/alexhernandezgarcia/gflownet/pull/79) - changes in the buffer to prevent an infinite loop.

* [wandb](https://wandb.ai/alexhg/public/runs/7x9d0vtq) of `python main.py gflownet=flowmatch env=grid proxy=corners`
* [wandb](https://wandb.ai/alexhg/public/runs/w0o1q8fl) of `python main.py gflownet=trajectorybalance env=grid proxy=corners`

### CODE REVIEW

File: gflownet/gflownet.py [hunk 1]
```diff
@@ -57,8 +57,16 @@ def __init__(
...
+            print(
+                """
+            Flow matching loss is not available for continuous environments.
+            Trajectory balance will be used instead
+            """
+            )
+            optimizer.loss = "tb"
         # Loss
-        if optimizer.loss in ["flowmatch"]:
+        if optimizer.loss in ["flowmatch", "flowmatching"]:
```
Comment (nikita-0209) [inline gflownet/gflownet.py:69]: I'm sorry, but what is `flowmatching`? I mean why would it be an option in the config when we already have `flowmatch`?

File: gflownet/gflownet.py [hunk 2]
```diff
@@ -524,68 +532,61 @@ def flowmatch_loss(self, it, batch, loginf=1000):
...
-        ]
-
-        # log(eps + exp(log(Q(s,a)))) : qsa
-        in_flow = torch.log(
-            self.loss_eps
-            + self._tfloat(torch.zeros((sp.shape[0],))).index_add_(
-                0, batch_idxs, torch.exp(parents_Qsa)
-            )
+        # Concatenate lists of tensors
+        states, parents, parents_a, done, masks_sf = map(
```
Comment (nikita-0209) [inline gflownet/gflownet.py:552]: I think we might have to deal with `states` and `parents` separately for variable length sequences as we did for `tb`

File: gflownet/gflownet.py (hunk 1 above)
Comment (alexhernandezgarcia) [reply gflownet/gflownet.py:69]: of this is just to allow for both names ("flowmatch" or "flowmatching"), like for the trajectory balance loss we allow both "trajectorybalance" and "tb". Maybe we should add "fm" to the flowmatch loss too :P

File: gflownet/gflownet.py (hunk 2 above)
Comment (alexhernandezgarcia) [reply gflownet/gflownet.py:552]: uhm but that actually broke a bunch of things :sweat_smile: let's deal with it when the need arises :grimacing:

File: gflownet/gflownet.py (hunk 1 above)
Comment (nikita-0209) [reply gflownet/gflownet.py:69]: lol okay

File: gflownet/gflownet.py (hunk 2 above)
Comment (nikita-0209) [reply gflownet/gflownet.py:552]: Yeah so in [PR#68](https://github.com/alexhernandezgarcia/gflownet/pull/68/files#diff-700050cd4007666a2c357edc47c03336bdbaddab3e135d60a3f382718580780b), I slightly modify that implementation by introducing a `do_padding` variable in the env, which is `False` by default (so that it does not break the grid/torus envs) and can be explicitly set to `True` for AMP

**Review by AlexandraVolokhova — APPROVED** [review]
looks good to me!

File: gflownet/gflownet.py (hunk 2 above)
Comment (alexhernandezgarcia) [reply gflownet/gflownet.py:552]: Ah that sounds better. I think we should merge this for now and you can add that part to the flowmatch loss if you ever need to use it?
